apns2/client.py

```diff
@@ -1,39 +1,149 @@
+        notification,
+        topic,
+        priority=NotificationPriority.Immediate
+    ):
+        stream_id = self.send_notification_async(token_hex, notification, topic, priority)
+        result = self.get_notification_result(stream_id)
+        if result != 'Success':
+            raise exception_class_for_reason(result)
 
-    def send_notification(self, token_hex, notification, priority=NotificationPriority.Immediate, topic=None):
-        json_payload = dumps(notification.dict(), ensure_ascii=False, separators=(',', ':')).encode('utf-8')
+    def send_notification_async(
+        self,
+        token_hex,
+        notification,
+        topic,
+        priority=NotificationPriority.Immediate
+    ):
+        json_payload = json.dumps(
+            notification.dict(),
+            ensure_ascii=False,
+            separators=(',', ':')
+        ).encode('utf-8')
 
-        headers = {
-            'apns-priority': priority.value
-        }
-        if topic:
-            headers['apns-topic'] = topic
+        headers = {'apns-topic': topic}
+        if priority != DEFAULT_APNS_PRIORITY:
+            headers['apns-priority'] = priority.value
 
         url = '/3/device/{}'.format(token_hex)
-        stream_id = self.__connection.request('POST', url, json_payload, headers)
-        resp = self.__connection.get_response(stream_id)
-        if resp.status != 200:
-            raw_data = resp.read().decode('utf-8')
+        stream_id = self._connection.request('POST', url, json_payload, headers)
+        return stream_id
+
+    def get_notification_result(self, stream_id):
+        response = self._connection.get_response(stream_id)
+        if response.status == 200:
+            return 'Success'
+        else:
+            raw_data = response.read().decode('utf-8')
             data = json.loads(raw_data)
-            raise exception_class_for_reason(data['reason'])
+            return data['reason']
+
+    def send_notification_batch(
+        self,
+        tokens,
+        notification,
+        topic,
+        priority=NotificationPriority.Immediate
+    ):
+        '''
+        Send a notification to a list of tokens in batch. Instead of sending a synchronous request
+        for each token, send multiple requests concurrently. This is done on the same connection,
+        using HTTP/2 streams (one request per stream).
+
+        APNs allows many streams simultaneously, but the number of streams can vary depending on
+        server load. This method reads the SETTINGS frame sent by the server to figure out the
+        maximum number of concurrent streams. Typically, APNs reports a maximum of 500.
+        '''
+        # Make sure we're connected to APNs, so that we receive and process the server's SETTINGS
+        # frame before starting to send notifications.
+        self._connection.connect()
+        result_counter = collections.Counter()
+        open_streams = collections.deque()
+        token_iterator = iter(tokens)
```

Comment:
I can't help wondering if you could refactor this to use a for loop over tokens, instead of an iterator object.  I find Python's iterator API kind of ugly, although nowhere near as ugly as the old Java/C++ iterator APIs good riddance to them.